### PR TITLE
[RFC][#55] - Setup redux in web-client

### PR DESCRIPTION
**Pull request by jpmarques66**
**Summary:** This PR will add basic redux setup

**Description:**
In this PR I add redux basic configuration with an example to provide an easier application bootstrap.
If you think the example doesn't make much sense we can delete it. I just think it will come in handy because of all of the typescript code :)

### CODE REVIEW

**Comment by jpmarques66** [comment]
> I prefer the implementation of redux using `hooks` instead of the "old way", imo its cleaner and more readable.
> 
> https://react-redux.js.org/api/hooks <- Tell me what you think
> 
> Not so many times we start projects from scratch and have this opportunities, I would advise to add `immer` to the "redux flow" (and not only but out of scope here).
> 
> It reduces a bit the redux boilerplate and makes the code even cleaner and readable.
> Check here -> https://immerjs.github.io/immer/docs/introduction
> 
> NOTE:
> The two suggestions above only apply if:
> 
> * Most of the team feels comfortable with it
> * It integrates fine with typescript (that most of you know I know very little and have been avoiding 😊 )
> 
> Leaving the aprove to @comoser

Thanks for the feedback @luisFilipePT. All good suggestions! I'm going to add them 😄

**Comment by comoser** [comment]
> I prefer the implementation of redux using `hooks` instead of the "old way", imo its cleaner and more readable.
> 
> https://react-redux.js.org/api/hooks <- Tell me what you think
> 
> Not so many times we start projects from scratch and have this opportunities, I would advise to add `immer` to the "redux flow" (and not only but out of scope here).
> 
> It reduces a bit the redux boilerplate and makes the code even cleaner and readable.
> Check here -> https://immerjs.github.io/immer/docs/introduction
> 
> NOTE:
> The two suggestions above only apply if:
> 
> * Most of the team feels comfortable with it
> * It integrates fine with typescript (that most of you know I know very little and have been avoiding 😊 )
> 
> Leaving the aprove to @comoser

These are indeed very good suggestions, feel free to implement them @jpmarques66 !

And thank you @luisFilipePT for the helpful review 👍

File: web-client/package.json [hunk 1]
```diff
@@ -10,9 +10,14 @@
     "@types/node": "^12.0.0",
     "@types/react": "^16.9.0",
     "@types/react-dom": "^16.9.0",
+    "@types/react-redux": "^7.1.7",
+    "immer": "^6.0.2",
+    "prop-types": "^15.7.2",
```
Comment (comoser) [inline web-client/package.json]: do we need prop types here?

File: web-client/src/App.tsx [hunk 2]
```diff
@@ -1,27 +1,21 @@
...
+import { Provider } from 'react-redux';
+import { enableAllPlugins } from "immer"
+import ExampleContainer from './containers/exampleContainer';
+import configureStore from './store';
+
+// Later we can check if we need all immer plugins
+enableAllPlugins()
 
 function App(): ReactElement {
+  const store = configureStore()
```
Comment (comoser) [inline web-client/src/App.tsx]: if ';' is not mandatory, we should add the rule to enforce it, or the other way around. Personally prefer to have it than not

File: web-client/src/store/index.ts [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import { combineReducers, createStore } from "redux";
+import { composeWithDevTools } from "redux-devtools-extension";
```
Comment (comoser) [inline web-client/src/store/index.ts]: imports with double quotes instead of single quotes here

File: web-client/src/App.tsx (hunk 2 above)
Comment (RafaelFerreiraTVD) [reply web-client/src/App.tsx]: Personally i also agree with @comoser

File: web-client/src/store/index.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+const rootReducer = combineReducers({ exampleReducer })
+
+export type AppState = ReturnType<typeof rootReducer>;
+
+export default function configureStore() {
+    // const middlewares = [thunkMiddleware];
+
+    const store = createStore(
+        rootReducer,
+        composeWithDevTools()
```
Comment (comoser) [inline web-client/src/store/index.ts]: we can already leverage the DEV env flag in order to turn this on, not sure on the key, but you can investigate it on the react docs

File: web-client/package.json (hunk 1 above)
Comment (comoser) [reply web-client/package.json]: `interface` and `type` from TS should totally replace the need for `propTypes`

File: web-client/src/store/index.ts (hunk 3 above)
Comment (jpmarques66) [reply web-client/src/store/index.ts]: @comoser I created an [`issue`](https://github.com/reach4help/reach4help/issues/71) to enforce some rules on eslint. Maybe I'll leave these coding-style related comments for that issue.

File: web-client/src/App.tsx (hunk 2 above)
Comment (jpmarques66) [reply web-client/src/App.tsx]: Same as https://github.com/reach4help/reach4help/pull/68#discussion_r396341297

File: web-client/src/store/index.ts (hunk 4 above)
Comment (jpmarques66) [reply web-client/src/store/index.ts]: Yes, you're right, I'll add that! Thanks

File: web-client/src/store/index.ts (hunk 3 above)
Comment (comoser) [reply web-client/src/store/index.ts]: makes sense! go for it

File: web-client/package.json (hunk 1 above)
Comment (jpmarques66) [reply web-client/package.json]: > do we need prop types here?

I moved prop-types for devDependencies. Nicely pointed out.

> `interface` and `type` from TS should totally replace the need for `propTypes`

I actually had this doubt as well. From what I read, typescript serves as a check at build time. Prop-types works in runtime. There's actually a utility to infer prop-types from typescript that may become very useful. So, for now, I'd say we keep both. If it becomes evidently useless we'll remove it.

**Pull request merged**